hello-vulkan/app/src/main/cpp/vk_triangle/hellovk.h

```diff
@@ -0,0 +1,1309 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <android/asset_manager.h>
+#include <android/log.h>
+#include <android/native_window.h>
+#include <android/native_window_jni.h>
+#include <assert.h>
+#include <vulkan/vulkan.h>
+
+#include <array>
+#include <fstream>
+#include <map>
+#include <optional>
+#include <set>
+#include <sstream>
+#include <string>
+#include <vector>
+
+/**
+ * HelloVK contains the core of Vulkan pipeline setup. It includes recording
+ * draw commands as well as screen clearing during the render pass.
+ *
+ * Please refer to: https://vulkan-tutorial.com/ for a gentle Vulkan
+ * introduction.
+ */
+
+namespace vkt {
+#define LOG_TAG "hellovkjni"
+#define LOGI(...) __android_log_print(ANDROID_LOG_INFO, LOG_TAG, __VA_ARGS__)
+#define LOGE(...) __android_log_print(ANDROID_LOG_ERROR, LOG_TAG, __VA_ARGS__)
+
+const int MAX_FRAMES_IN_FLIGHT = 2;
+
+struct UniformBufferObject {
+  std::array<float, 16> mvp;
+};
+
+struct QueueFamilyIndices {
+  std::optional<uint32_t> graphicsFamily;
+  std::optional<uint32_t> presentFamily;
+  bool isComplete() {
+    return graphicsFamily.has_value() && presentFamily.has_value();
+  }
+};
+
+struct SwapChainSupportDetails {
+  VkSurfaceCapabilitiesKHR capabilities;
+  std::vector<VkSurfaceFormatKHR> formats;
+  std::vector<VkPresentModeKHR> presentModes;
+};
+
+struct ANativeWindowDeleter {
+  void operator()(ANativeWindow *window) { ANativeWindow_release(window); }
+};
+
+std::vector<uint8_t> LoadBinaryFileToVector(const char *file_path,
+                                            AAssetManager *assetManager) {
+  std::vector<uint8_t> file_content;
+  assert(assetManager);
+  AAsset *file =
+      AAssetManager_open(assetManager, file_path, AASSET_MODE_BUFFER);
+  size_t file_length = AAsset_getLength(file);
+
+  file_content.resize(file_length);
+
+  AAsset_read(file, file_content.data(), file_length);
+  AAsset_close(file);
+  return file_content;
+}
+
+const char *
+to_string_message_severity(VkDebugUtilsMessageSeverityFlagBitsEXT s) {
+  switch (s) {
+  case VK_DEBUG_UTILS_MESSAGE_SEVERITY_VERBOSE_BIT_EXT:
+    return "VERBOSE";
+  case VK_DEBUG_UTILS_MESSAGE_SEVERITY_ERROR_BIT_EXT:
+    return "ERROR";
+  case VK_DEBUG_UTILS_MESSAGE_SEVERITY_WARNING_BIT_EXT:
+    return "WARNING";
+  case VK_DEBUG_UTILS_MESSAGE_SEVERITY_INFO_BIT_EXT:
+    return "INFO";
+  default:
+    return "UNKNOWN";
+  }
+}
+const char *to_string_message_type(VkDebugUtilsMessageTypeFlagsEXT s) {
+  if (s == 7)
+    return "General | Validation | Performance";
+  if (s == 6)
+    return "Validation | Performance";
+  if (s == 5)
+    return "General | Performance";
+  if (s == 4 /*VK_DEBUG_UTILS_MESSAGE_TYPE_PERFORMANCE_BIT_EXT*/)
```

Comment:
why don't we use vk enums directly here?

Reply:
I've copied that from the vk bootstrap library. I imagine there's a reason they haven't used it. I can look into it if you think it's important.

Reply:
and you're using the very same enums in populateDebugMessengerCreateInfo function...

Reply:
ah, I see what you mean now! misinterpreted your comment.

👍